### PR TITLE
chore: update deepspeed launcher to work with newer versions

### PR DESCRIPTION
**Pull request by liamcli**
## Description
There have been changes to the deepspeed launcher in newer versions that break compatibility with our launcher.  This PR addresses those compatibility issues introduced in 

- https://github.com/microsoft/DeepSpeed/pull/1832
- https://github.com/microsoft/DeepSpeed/pull/1907

Using deepspeed versions > 0.6.4 will depend on this upstream PR https://github.com/microsoft/DeepSpeed/pull/1957

## Test Plan
Tested single and multinode for 0.5.10 and 0.6.5.dev with the PR above

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/determined/launch/deepspeed.py [hunk 1]
```diff
@@ -7,34 +7,38 @@
...
     # Instead, the deepspeed launcher will use localhost as the chief worker ip.
     if len(ip_addresses) == 1:
         trial_runner_hosts[0] = "localhost"
 
-    os.makedirs(hostfile_path.parent, exist_ok=True)
-    with open(hostfile_path, "w") as hostfile:
-        lines = [f"{host} slots={num_proc_per_machine}\n" for host in trial_runner_hosts]
-        hostfile.writelines(lines)
+    if hostfile_path is not None:
+        os.makedirs(pathlib.Path(hostfile_path).parent, exist_ok=True)
```
Comment (rb-determined-ai) [inline harness/determined/launch/deepspeed.py:38]: nit: stick to either str or pathlib, don't go back and forth.

Replace `pathlib.Path(hostfile_path).parent` with `os.path.dirname(hostfile_path)`

File: harness/determined/launch/deepspeed.py [hunk 2]
```diff
@@ -100,29 +104,42 @@ def create_deepspeed_env_file() -> None:
         environ = os.environ.copy()
         for k, v in environ.items():
             if k in INCLUDE:
-                f.write(f"{k}={v}\n")
+                if deepspeed_version >= version.parse("0.6.2"):
```
Comment (rb-determined-ai) [inline harness/determined/launch/deepspeed.py:107]: when you introduce a version check like this, can you include a comment explaining the difference and motivation?

File: harness/determined/launch/deepspeed.py [hunk 3]
```diff
@@ -100,29 +104,42 @@ def create_deepspeed_env_file() -> None:
...
-        "--no_python",
-        "--no_local_rank",
-        "--",
-    ]
+    deepspeed_process_cmd = ["deepspeed"]
+    if hostfile_path is not None:
+        deepspeed_process_cmd += ["-H", hostfile_path]
+    deepspeed_process_cmd += ["--master_addr", master_address, "--no_python", "--no_local_rank"]
+    if deepspeed_version > version.parse("0.6.4"):
+        deepspeed_process_cmd.append("--no_ssh_check")
```
Comment (rb-determined-ai) [inline harness/determined/launch/deepspeed.py:120]: question: how did we avoid the need for this flag before?

File: harness/determined/launch/deepspeed.py [hunk 4]
```diff
@@ -100,29 +104,42 @@ def create_deepspeed_env_file() -> None:
...
 
 
+def check_deepspeed_version(multi_machine: bool) -> None:
+    if multi_machine:
+        compatible = True
+        if deepspeed_version > version.parse("0.6.0") and deepspeed_version <= version.parse(
+            "0.6.4"
+        ):
+            compatible = False
+        assert compatible, f"launcher incompatible with deepspeed version {deepspeed.__version__}"
```
Comment (rb-determined-ai) [inline harness/determined/launch/deepspeed.py:132]: nit: simply the logic/indentation here:

```python
def check_deepspeed_version(multi_machine: bool) -> None:
    if not multi_machine:
        return
    if deepspeed_version <= version.parse("0.6.0"):
        return
    if deepspeed_version > version.parse("0.6.4"):
        return
    raise ValueError(...)
```

Also, expand the error message to explain which versions _are_ allowed (so the user knows what to do) and explain why the disallowed versions are disallowed (so the user understands why we can't support them).

**Review by rb-determined-ai — COMMENTED** [review]
question: is there any benefit to landing this before an upstream release with `--no_ssh_check` is cut?

File: harness/determined/launch/deepspeed.py (hunk 3 above)
Comment (liamcli) [reply harness/determined/launch/deepspeed.py:120]: one of the PRs linked above recently added ssh check to the deepspeed launcher but it doesn't have all the flags we require for ssh (i.e. StrictHostKeyChecking=no)

**Comment by liamcli** [comment]
> question: is there any benefit to landing this before an upstream release with `--no_ssh_check` is cut?

i think better to wait.  i'll also update our environment image once that is released.

**Comment by rb-determined-ai** [comment]
@liamcli is [now](https://pypi.org/project/deepspeed/#history) good?

File: harness/determined/launch/deepspeed.py [hunk 5]
```diff
@@ -153,28 +162,53 @@ def create_deepspeed_env_file() -> None:
...
+    deepspeed_process_cmd.append("--")
     return deepspeed_process_cmd
 
 
+def check_deepspeed_version(multi_machine: bool) -> None:
+    if not multi_machine:
+        return
+    # Upstream deepspeed added an ssh check from 0.6.1 onwards that does not have the
+    # StrictHostKeyChecking=no arg and fails with our agents.  A PR adding a no_ssh_check arg
+    # to bypass this should land for versions 0.6.5 and onwards.
```
Comment (rb-determined-ai) [inline harness/determined/launch/deepspeed.py:191]: it already did land, right?

**Review by rb-determined-ai — APPROVED** [review]
nice

File: harness/determined/launch/deepspeed.py (hunk 5 above)
Comment (liamcli) [reply harness/determined/launch/deepspeed.py:191]: yes, latest version will work with our code

**Pull request merged**